src/core/halide2isl.cc

```diff
@@ -82,44 +82,141 @@ isl::aff makeIslAffFromInt(isl::space space, int64_t val) {
+       makeIslAffBoundsFromExpr(space, op->a, allowMin, allowMax)) {
+    result.push_back(aff);
+  }
+  for (const auto& aff :
+       makeIslAffBoundsFromExpr(space, op->b, allowMin, allowMax)) {
+    result.push_back(aff);
+  }
+
+  return result;
+}
+
+/*
+ * Convert Halide binary expression "op" into an isl affine function by
+ * converting its LHS and RHS into affs and combining them with "combine"
+ * into a single expression.  LHS and RHS are expected to only produce a single
+ * expression.
+ * This is intended for use with operations other than Min/Max that do not
+ * commute nicely in bounds, for example
+ *   x < a + max(b,c)  NOT <=>  x < a + b AND x < a + c for negative values.
+ */
+template <typename T>
+inline isl::aff combineSingleAffs(
+    isl::space space,
+    T op,
+    isl::aff (isl::aff::*combine)(isl::aff) const) {
+  auto left = makeIslAffBoundsFromExpr(space, op->a, false, false);
+  auto right = makeIslAffBoundsFromExpr(space, op->b, false, false);
+  CHECK_EQ(left.size(), 1);
+  CHECK_EQ(right.size(), 1);
+
+  return (left[0].*combine)(right[0]);
+}
+
+} // end namespace
+
+/*
+ * Convert Halide expression into list of isl affine expressions usable for
+ * defining constraints.  In particular, an expression starting with (nested)
+ * Max operations can be used for lower bounds
+ *   x > max(a,b) <=> x > a AND x > b
+ * while an expression starting with (nested) Min operations can be used for
+ * upper bounds
+ *   x < min(a,b) <=> x < a AND x < b.
+ * Arguments "allowMin" and "allowMax" control whether Min and Max operations,
+ * respecitvely, are allowed to be present in the expression. Note that they
+ * can only appear before any other operation and cannot appear together in an
+ * expression.
+ * If a Halide expression cannot be converted into a list of affine expressions,
+ * return an empty list.
+ */
+std::vector<isl::aff> makeIslAffBoundsFromExpr(
+    isl::space space,
+    const Expr& e,
+    bool allowMin,
+    bool allowMax) {
+  CHECK(!(allowMin && allowMax));
+
+  const Min* minOp = e.as<Min>();
+  const Max* maxOp = e.as<Max>();
+
   if (const Variable* op = e.as<Variable>()) {
     isl::local_space ls = isl::local_space(space);
     int pos = space.find_dim_by_name(isl::dim_type::param, op->name);
     if (pos >= 0) {
-      return isl::aff(ls, isl::dim_type::param, pos);
+      return {isl::aff(ls, isl::dim_type::param, pos)};
     } else {
+      // FIXME: thou shalt not rely upon set dimension names
       pos = space.find_dim_by_name(isl::dim_type::set, op->name);
       if (pos >= 0) {
-        return isl::aff(ls, isl::dim_type::set, pos);
+        return {isl::aff(ls, isl::dim_type::set, pos)};
       }
     }
     LOG(FATAL) << "Variable not found in isl::space: " << space << ": " << op
                << ": " << op->name << '\n';
-    return isl::aff();
+    return {};
+  } else if (minOp != nullptr && allowMin) {
+    return concatAffs(space, minOp, allowMin, allowMax);
```

Comment:
Once you've allowed a min, should allowMax still be true? What happens with x < max(min(max(min(....))))

Reply:
The trick with multiple constraints does not work, so it should not be allowed.  There's a check above that allowMin and allowMax should not be true at the same time.

Reply:
Ah, didn't see that check